tests/chainer_tests/iterators_tests/test_multiprocess_iterator.py

```diff
@@ -374,4 +391,162 @@ def test_iterator_serialize_backward_compat(self):
         self.assertAlmostEqual(it.previous_epoch_detail, 4 / 6)
 
 
+class TestMultiprocessIteratorConcurrency(unittest.TestCase):
+
+    def test_finalize_not_deadlock(self):
+        dataset = numpy.ones((1000, 1000))
+        it = iterators.MultiprocessIterator(dataset, 10, n_processes=4)
+        for _ in range(10):
+            it.next()
+
+        timeout = 5.
+        t = threading.Thread(target=lambda: it.finalize())
+        t.daemon = True
+        t.start()
+        t.join(timeout)
+        deadlock = t.is_alive()
+
+        self.assertFalse(deadlock)
+
+
+class TestMultiprocessIteratorInterruption(unittest.TestCase):
+
+    # unless you're debugging tests, this should be false
+    show_interruption_msg = False
+
+    def setUp(self):
+        self.code_path = None
+        self.sys_path_appended = False
+        if not self.__class__.show_interruption_msg:
+            self.nullfd = os.open(os.devnull, os.O_WRONLY)
+
+    def tearDown(self):
+        if not self.__class__.show_interruption_msg:
+            os.close(self.nullfd)
+        if self.code_path is not None:
+            os.remove(self.code_path)
+        if self.sys_path_appended:
+            sys.path.pop()
+
+    def run_code(self, dataset, n_processes, n_prefetch, shared_mem,
+                 operation):
+        code_template = """
+import random
+import time
+from chainer import iterators
+
+class InfiniteWaitDataSet(object):
+    def __len__(self):
+        return 1000000
+    def __getitem__(self, _):
+        time.sleep(1000000)
+infinite_wait = InfiniteWaitDataSet()
+
+class NoWaitDataSet(object):
+    def __len__(self):
+        return 1000000
+    def __getitem__(self, _):
+        return 0
+no_wait = NoWaitDataSet()
+
+if __name__ == '__main__':
+    if {shared_mem} is not None and {dataset} is infinite_wait:
+        iterators.MultiprocessIterator._interruption_testing = True
+    it = iterators.MultiprocessIterator({dataset}, 100,
+                                        n_processes={n_processes},
+                                        n_prefetch={n_prefetch},
+                                        shared_mem={shared_mem})
+    {operation}
+        """
+        code = code_template.format(dataset=dataset,
+                                    n_processes=n_processes,
+                                    n_prefetch=n_prefetch,
+                                    shared_mem=shared_mem,
+                                    operation=operation)
+        fd, self.code_path = tempfile.mkstemp(suffix='.py')
+        os.write(fd, six.b(code))
+        os.close(fd)
+        sys.path.append(os.path.dirname(self.code_path))
```

Comment:
Please use absolute path instead this workaround.

Reply:
Whether do you mean?

1. instead of dynamically creating a python script using template string, use statically created python script; or
2. use absolute path for the dynamically created script, not the one created by `tempfile.mkstemp`.

If you mean (1), it seems somehow reasonable, but I don't known where I should create the python script.

I don't agree with (2). Temporal files should be on the standard location. But if the chainer already uses some fixed temporal directory, I won't oppose to use it.

Reply:
I confused you.
I think that it is possible to run a dynamically generated file using an absolute path.
Maybe we do not modify `sys.path`.